File: src/types/financial.ts
export type CostCoverageStatus = 'BELOW_TYPICAL_RANGE' | 'WITHIN_TYPICAL_RANGE' | 'ABOVE_TYPICAL_RANGE'

export type FinancialReadinessStatus =
  | 'INSUFFICIENT_CAPITAL_STRUCTURE'
  | 'PARTIALLY_ALIGNED'
  | 'ALIGNED'
  | 'ABOVE_TYPICAL_REQUIREMENT'

export interface FinancialBusiness {
  id: string
  slug: string
  name: string
  category: string
  shortDescription: string
}

export interface BusinessCostSnapshot {
  minimumCapital: string
  maximumCapital: string
  setupCostMin: string
  setupCostMax: string
  estimatedSetupCostMin?: string
  estimatedSetupCostMax?: string
  workingCapitalMin: string
  workingCapitalMax: string
}

export interface BusinessCostAlignment {
  setupCostCoverageStatus: CostCoverageStatus
  financialReadinessStatus: FinancialReadinessStatus
  fundingGap: string
  capacityAboveEstimatedMax: string
  comparisonExplanation: string
  workingCapitalExplanation: string
}

export interface FinancialWarning {
  code: string
  message: string
}

export interface FinancialAnalysis {
  id: string
  createdAt: string
  analysisVersion: string
  feasibilityAnalysisId?: string
  business: FinancialBusiness
  availableMarginCapital: string
  beneficiaryContribution: string
  beneficiaryContributionPercentage?: string
  feasibleProjectCost: string
  indicativeLoanAmount: string
  indicativeLoanPercentage?: string
  businessCosts: BusinessCostSnapshot
  alignment: BusinessCostAlignment
  warnings: FinancialWarning[]
  disclaimer: string
  dataSourceNotice: string
  nextStep: string
}

File: backend/app/schemas/financial.py
import uuid
from datetime import datetime
from decimal import Decimal, InvalidOperation
from typing import List, Optional

from pydantic import BaseModel, Field, field_validator

from app.financial_rules import MAXIMUM_MARGIN_INPUT


class FinancialAnalysisRequest(BaseModel):
    business_slug: str = Field(min_length=1, max_length=100)
    available_margin_capital: Decimal
    feasibility_analysis_id: Optional[uuid.UUID] = None

    @field_validator("available_margin_capital", mode="before")
    @classmethod
    def exact_margin(cls, value):
        if isinstance(value, (float, bool)):
            raise ValueError("Available margin capital must be sent as a decimal string or integer, not a float.")
        try:
            amount = Decimal(value)
        except (InvalidOperation, TypeError, ValueError) as exc:
            raise ValueError("Enter a valid available margin capital amount.") from exc
        if not amount.is_finite() or amount <= 0:
            raise ValueError("Available margin capital must be greater than zero.")
        if amount.as_tuple().exponent < -2:
            raise ValueError("Available margin capital supports at most two decimal places.")
        if amount > MAXIMUM_MARGIN_INPUT:
            raise ValueError("Available margin capital exceeds the ₹1 crore prototype input limit.")
        return amount.quantize(Decimal("0.01"))


class FinancialBusiness(BaseModel):
    id: uuid.UUID
    slug: str
    name: str
    category: str
    short_description: str


class FinancialWarning(BaseModel):
    code: str
    message: str


class BusinessCostAlignment(BaseModel):
    setup_cost_coverage_status: str
    financial_readiness_status: str
    funding_gap: str
    capacity_above_estimated_max: str
    comparison_explanation: str
    working_capital_explanation: str


class FinancialAnalysisResponse(BaseModel):
    id: uuid.UUID
    created_at: datetime
    analysis_version: str
    feasibility_analysis_id: Optional[uuid.UUID]
    business: FinancialBusiness
    available_margin_capital: str
    beneficiary_contribution: str
    beneficiary_contribution_percentage: str = "10.00"
    feasible_project_cost: str
    indicative_loan_amount: str
    indicative_loan_percentage: str = "90.00"
    business_costs: dict
    alignment: BusinessCostAlignment
    warnings: List[FinancialWarning]
    disclaimer: str
    data_source_notice: str
    next_step: str
